lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java

```diff
@@ -25,9 +36,9 @@ public static Parser create() {
         var state = allocState();
         return new Parser(state);
     }
-    public final Tree parse(String input) throws UnsupportedSyntaxException {
+    public final Tree parse(CharSequence input) throws UnsupportedSyntaxException {
```

Comment:
The question is how huge the underlying _direct_ `ByteBuffer` is supposed to be? Can we overallocate? Is the buffer just temporary? Or shall we try to allocate as small `ByteBuffer` as possible?

There is [CharsetEncoder](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/charset/CharsetEncoder.html) and it has methods like `averageBytesPerChar()` or [maxBytesPerChar()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/charset/CharsetEncoder.html#maxBytesPerChar()). We can use those methods to estimate...

Other useful method may be [wrap](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/CharBuffer.html#wrap(java.lang.CharSequence)) to convert `CharSequence` to `CharBuffer`.

Reply:
Oh, I thought we could get a `CharBuffer` out of the `Source`, but I was wrong.

Given that the `Tree`s/`Token`s don't need to be able to produce `String` representations from their fields alone, the parser can own and reuse a direct-allocated `CharBuffer`; it can accept a `CharSequence` as input, copy all the chars into the buffer, and loan the data to Rust just for the duration of the JNI call.